src/comb_filter.rs

```diff
@@ -21,26 +28,345 @@ pub enum Error {
+        let filter_coefficients = vec![0.0; delay_len];
+        let feedback = 0.5;
+
+        CombFilter {
+            delay_len,
+            delay_line,
+            filter_coefficients,
+            sample_rate_hz,
+            num_channels,
+            feedback, // default feedback value
+            filter_type,
+        }
     }
 
     pub fn reset(&mut self) {
-        todo!("implement")
+        // Zeroed delay line
+        for channel in &mut self.delay_line {
+            for sample in channel {
+                *sample = 0.0;
+            }
+        }
+        // Reset mutable states
+        self.feedback = 0.5;
     }
 
     pub fn process(&mut self, input: &[&[f32]], output: &mut [&mut [f32]]) {
-        todo!("implement");
+        match self.filter_type {
+            FilterType::FIR => {
+                // Matlab FIR Implementation
+                for (channel, & in_channel) in input.iter().enumerate() {
+                    let fir_delay_length = self.delay_line[channel].len();
+                    let filter_coeff_length = self.filter_coefficients.len();
+
+                    for (samp, &in_samp) in in_channel.iter().enumerate() {
+                        let mut out_del_sample = 0.0;
+
+                        // FIR difference equation
+                        for n in 0..filter_coeff_length {
+                            if samp >= n {
+                                out_del_sample += in_channel[samp - n] * (self.filter_coefficients[n] as f32);
+                            }
+                        }
+                        output[channel][samp] = out_del_sample;
+
+                        // delay line update
+                        let mut update_delay_line = vec![0.0; fir_delay_length];
+                        update_delay_line[0] = in_samp;
+                        update_delay_line[1..].copy_from_slice(&self.delay_line[channel][0..(fir_delay_length - 1)]);
+                        self.delay_line[channel] = update_delay_line;
+                    }
+                }
+            }
+            FilterType::IIR => {
+                // Matlab IIR Implementation
+                for (channel, & in_channel) in input.iter().enumerate() {
+                    let iir_delay_length = self.delay_line[channel].len();
+
+                    for (samp, &in_samp) in in_channel.iter().enumerate() {
+                        // IIR difference equation
+                        let out_del_sample = self.delay_line[channel][iir_delay_length - 1] * self.feedback;
+                        output[channel][samp] = in_samp + out_del_sample;
+
+                        // delay line update
+                        let mut update_delay_line = vec![0.0; iir_delay_length];
+                        update_delay_line[0] = output[channel][samp];
+                        update_delay_line[1..].copy_from_slice(&self.delay_line[channel][0..(iir_delay_length - 1)]);
+                        self.delay_line[channel] = update_delay_line;
```

Comment:
If you are trying to shift every elements one to the right, and put the calculation at the first index. You can try using the ```self.delay_line.rotate_right(1)``` and ```self.delay_line[0] = output[channel][samp]```. That way, you can save some lines of codes